streams/src/test/java/org/apache/kafka/streams/integration/KTableEfficientRangeQueryTest.java

```diff
@@ -0,0 +1,259 @@
+            }
+            if (to != null && k.compareTo(to) > 0) {
+                continue;
+            }
+            result.add(record);
+        }
+        return result;
+    }
+
+    private void testRange(final String name, final ReadOnlyKeyValueStore<String, String> store, final String from, final String to, final boolean forward) {
+        final KeyValueIterator<String, String> resultIterator = forward ? store.range(from, to) : store.reverseRange(from, to);
+        final List<KeyValue<String, String>> result = getOrderedContents(resultIterator);
+        final KeyValueIterator<String, String> expectedIterator = forward ? store.all() : store.reverseAll();
+        final List<KeyValue<String, String>> expected = filterList(expectedIterator, from, to);
+        assertThat(result, is(expected));
+    }
+
+    private static Collection<Object[]> buildParameters(final List<?>... argOptions) {
+        List<Object[]> result = new LinkedList<>();
+        result.add(new Object[0]);
+
+        for (final List<?> argOption : argOptions) {
+            result = times(result, argOption);
+        }
+
+        return result;
+    }
+
+    private static List<Object[]> times(final List<Object[]> left, final List<?> right) {
+        final List<Object[]> result = new LinkedList<>();
+        for (final Object[] args : left) {
+            for (final Object rightElem : right) {
+                final Object[] resArgs = new Object[args.length + 1];
+                System.arraycopy(args, 0, resArgs, 0, args.length);
+                resArgs[args.length] = rightElem;
+                result.add(resArgs);
+            }
+        }
+        return result;
+    }
+
+    private Materialized<String, String, KeyValueStore<Bytes, byte[]>> getStoreConfig(final StoreType type, final String name, final boolean cachingEnabled, final boolean loggingEnabled) {
+        final Supplier<KeyValueBytesStoreSupplier> createStore = () -> {
+            if (type == StoreType.InMemory) {
+                return Stores.inMemoryKeyValueStore(TABLE_NAME);
+            } else if (type == StoreType.RocksDB) {
+                return Stores.persistentKeyValueStore(TABLE_NAME);
+            } else if (type == StoreType.Timed) {
+                return Stores.persistentTimestampedKeyValueStore(TABLE_NAME);
+            } else {
+                return Stores.inMemoryKeyValueStore(TABLE_NAME);
+            }
+        };
+
+        final KeyValueBytesStoreSupplier stateStoreSupplier = createStore.get();
+        Materialized<String, String, KeyValueStore<Bytes, byte[]>> stateStoreConfig = Materialized
+                .<String, String>as(stateStoreSupplier)
+                .withKeySerde(Serdes.String())
+                .withValueSerde(Serdes.String());
+        if (!cachingEnabled) {
+            stateStoreConfig = stateStoreConfig.withCachingDisabled();
+        }
+        if (!loggingEnabled) {
+            stateStoreConfig = stateStoreConfig.withLoggingDisabled();
+        }
```

Comment:
Do you think it would be clearer if we don't rely on the defaults, but just explicitly include both branches?

Oh, also, this isn't an immutable builder, so you can just do:
```suggestion
        if (cachingEnabled) {
            stateStoreConfig.withCachingEnabled();
        } else {
            stateStoreConfig.withCachingDisabled();
        }
        if (loggingEnabled) {
            stateStoreConfig.withLoggingEnabled(new HashMap());
        } else {
            stateStoreConfig.withLoggingDisabled();
        }
```

Reply:
Makes sense!